src/sync/mpsc.rs

```diff
@@ -0,0 +1,134 @@
+//! A multi-producer, single-consumer, futures-aware channel
+
+use {Async, Poll};
+use std::sync::atomic::Ordering::SeqCst;
+use std::sync::atomic::AtomicUsize;
+use stream::Stream;
+use task::{self, Task};
+use std::sync::{Arc, mpsc};
+use lock::Lock;
+
+/// `Sender` of the channel. `Sender` can be cloned.
+pub struct Sender<T> {
+    /// Sender
+    tx: mpsc::Sender<T>,
+    /// Inner
+    inner: Arc<Inner>,
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Sender<T> {
+        let _ = self.inner.sender_count.fetch_add(1, SeqCst);
+        Sender {
+            tx: self.tx.clone(),
+            inner: self.inner.clone(),
+        }
+    }
+}
+
+/// Shared state across the `Sender`(s) and `Receiver`.
+struct Inner {
+    /// Stores the `Receiver` task. Only one receiver task is used.
+    rx_task: Lock<Option<Task>>,
+    /// Count of senders, needed for determining when the last sender has been dropped.
+    sender_count: AtomicUsize,
+}
+
+/// Receiver for the channel.
+#[must_use = "streams do nothing unless polled"]
+pub struct Receiver<T> {
+    /// Receiver
+    rx: mpsc::Receiver<T>,
+    /// Inner
+    inner: Arc<Inner>,
+}
+
+impl<T> Sender<T> {
+    /// Send a message on the channel
+    pub fn send(&self, t: T) -> Result<(), mpsc::SendError<T>> {
+        match self.tx.send(t) {
+            Ok(_) => {
+                // if an rx_task exists, try to unpark
+                // otherwise someone else is unparking the Receiver
+                if let Some(mut opt_task) = self.inner.rx_task.try_lock() {
+                    if let Some(task) = opt_task.take() {
+                        task.unpark();
```

Comment:
Ah I tend to try to be careful to do operations like this *outside* of the lock, so could this be accompanied with a `drop(opt_task)` before?